### PR TITLE
Add an AsyncImage widget

### PR DESCRIPTION
**Pull request by donadigo**
I've created this widget to address issues when showing UI's with lots of icons like in Slingshot so that we could use it when appropriate. When we switch to this, things will be a little bit more responsive for slower hard drives. For more details on how the widget works, look at it's documentation and the included demo.

### CODE REVIEW

**Comment by donadigo** [comment]
Branch ready to review again.

File: lib/Widgets/AsyncImage.vala [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+* License along with this program; if not, write to the
+* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+* Boston, MA 02110-1301 USA
+*/
+
+/**
+ * AsyncImage is an {@link Gtk.Image} that provides a way to load
+ * icons and images asynchronously without blocking the main GTK thread.
+ * 
+ * AsyncImage can be used to improve your's GTK interface performance that
```
Comment (Philip-Scott) [inline lib/Widgets/AsyncImage.vala]: I think it should be` to improve your GTK interface's`

File: lib/Widgets/AsyncImage.vala [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+ * 
+ * Even though AsyncImage sets only the {@link Gtk.Image.surface}, it automatically detects changes to the underlying {@link Gtk.Widget.scale_factor}
+ * and reloads the icon to a new scale factor when it changes. If you request to set an {@link GLib.ThemedIcon} and the icon or GTK theme changes
+ * the AsyncImage will also reload it to display the new icon with applied changes.
+ * 
+ * The {@link Granite.AsyncImage.gicon_async} and {@link Granite.AsyncImage.size_async} are properties which reflect
+ * the current icon and it's size which will or is currently displayed. Note that those two properties will return
+ * meaningful results //''only''// when you call {@link Granite.AsyncImage.set_from_gicon_async} and it's wrappers.
+ * 
+ * AsyncImage has also it's own cache for already loaded icons. If you attempt to load the same icon at the same size
```
Comment (Philip-Scott) [inline lib/Widgets/AsyncImage.vala]: `its own cache` :)

File: lib/Widgets/AsyncImage.vala [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+
+        if (!load_on_realize) {
+            yield set_from_gicon_async_internal (gicon_async, size_async, cancellable, false);
+        }
+    }
+
+    /**
+     * A wrapper for {@link Granite.AsyncImage.set_from_gicon_async} to display an icon name.
+     * 
+     * This is a convinience method for setting an icon name with a desired {@link Gtk.IconSize}. Note that you'll not be
```
Comment (Philip-Scott) [inline lib/Widgets/AsyncImage.vala]: convinience -> convenience

**Comment by Philip-Scott** [comment]
Also made a comment about the functionality in elementary/applications-menu#48 :)

File: lib/Widgets/AsyncImage.vala [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+
+    /**
+     * If the widget should act as a placeholder when the image is not yet loaded.
+     * 
+     * For more details see {@link Granite.AsyncImage.AsyncImage}.
+     */
+    public bool auto_size_request { construct; private get; }
+
+    /**
+     * The current icon that will or is displayed in the image.
```
Comment (danrabbit) [inline lib/Widgets/AsyncImage.vala]: "The icon that will be or is currently displayed in the image"

**Pull request merged**